org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -0,0 +1,169 @@
+package org.code.javabuilder;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.security.Principal;
+import java.util.Arrays;
+import javax.tools.Diagnostic;
+import javax.tools.DiagnosticCollector;
+import javax.tools.JavaCompiler;
+import javax.tools.JavaCompiler.CompilationTask;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.StandardLocation;
+import javax.tools.ToolProvider;
+
+public class JavaRunner {
+
+  /**
+   * Compile and Run the given user program and output results to the given compileRunService for
+   * the given user (principal)
+   *
+   * @param userProgram
+   * @param principal
+   * @param compileRunService
+   */
+  public static void compileAndRunUserProgram(
+      UserProgram userProgram, Principal principal, CompileRunService compileRunService) {
+    String filename = userProgram.getFileName();
+    // We expect the filename to have no .java suffix, remove it if necessary.
+    if (filename.endsWith(".java")) {
+      userProgram.setFileName(filename.substring(0, filename.indexOf(".java")));
+    }
+
+    File tempFolder = null;
+    try {
+      tempFolder = Files.createTempDirectory("tmpdir").toFile();
+
+      compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+      boolean compileSuccess =
+          compileProgram(userProgram, principal, compileRunService, tempFolder);
+
+      if (compileSuccess) {
+        // set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        System.setOut(out);
+
+        compileRunService.sendMessages(principal.getName(), "Compiled!");
+        compileRunService.sendMessages(principal.getName(), "Running your program...");
```

Comment:
Definitely out of scope for this PR, but I wanted to mention it: We'll want to eventually consider how we internationalize strings from the JavaBuilder service. Or whether we can find a way to avoid using strings in the JavaBuilder service.